File: backend/app/schemas/airport.py
"""pydantic schemas for airport endpoints."""

from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field, computed_field

from app.schemas.common import ListMeta
from app.schemas.geometry import PointZ
from app.schemas.infrastructure import (
    ObstacleResponse,
    SafetyZoneResponse,
    SurfaceResponse,
)


class AirportCreate(BaseModel):
    """airport create schema."""

    icao_code: str = Field(min_length=4, max_length=4, pattern=r"^[A-Z]{4}$")
    name: str
    city: str | None = None
    country: str | None = None
    elevation: float
    location: PointZ


class AirportUpdate(BaseModel):
    """airport update schema."""

    name: str | None = None
    city: str | None = None
    country: str | None = None
    elevation: float | None = None
    location: PointZ | None = None


class AirportResponse(BaseModel):
    """airport response schema."""

    id: UUID
    icao_code: str
    name: str
    city: str | None = None
    country: str | None = None
    elevation: float
    location: PointZ
    default_drone_profile_id: UUID | None = None
    terrain_source: Literal["FLAT", "DEM_UPLOAD", "DEM_API", "DEM_SRTM"] = "FLAT"
    dem_file_path: str | None = Field(default=None, exclude=True)

    @computed_field
    @property
    def has_dem(self) -> bool:
        """whether the airport has a DEM file configured."""
        return self.dem_file_path is not None

    model_config = {"from_attributes": True}


class SetDefaultDroneRequest(BaseModel):
    """request to set or clear the default drone for an airport."""

    drone_profile_id: UUID | None = None


class BulkChangeDroneRequest(BaseModel):
    """request to bulk-change drone profile on missions."""

    drone_profile_id: UUID
    from_drone_id: UUID | None = None
    scope: Literal["ALL_DRAFT", "SELECTED"] = "ALL_DRAFT"
    mission_ids: list[UUID] = []


class BulkChangeDroneResponse(BaseModel):
    """response for bulk drone change operation."""

    updated_count: int
    regressed_count: int = 0
    mission_ids: list[UUID]


class AirportDetailResponse(AirportResponse):
    """airport detail response schema."""

    surfaces: list[SurfaceResponse] = []
    obstacles: list[ObstacleResponse] = []
    safety_zones: list[SafetyZoneResponse] = []


class AirportSummaryResponse(AirportResponse):
    """airport with infrastructure counts for the selection table."""

    surfaces_count: int = 0
    agls_count: int = 0
    missions_count: int = 0


class AirportListResponse(BaseModel):
    """airport list response schema."""

    data: list[AirportResponse]
    meta: ListMeta


class AirportSummaryListResponse(BaseModel):
    """airport summary list response schema."""

    data: list[AirportSummaryResponse]
    meta: ListMeta


class TerrainCoverage(BaseModel):
    """terrain DEM coverage info."""

    bounds: tuple[float, float, float, float]
    resolution: tuple[float, float]


class TerrainUploadResponse(BaseModel):
    """response after uploading a DEM file."""

    terrain_source: Literal["FLAT", "DEM_UPLOAD", "DEM_API"]
    coverage: TerrainCoverage


class TerrainDownloadResponse(BaseModel):
    """response after downloading elevation data from API."""

    terrain_source: Literal["FLAT", "DEM_UPLOAD", "DEM_API"]
    points_downloaded: int
    coverage: TerrainCoverage


class ElevationAtPointResponse(BaseModel):
    """ground elevation at a single (lat, lon)."""

    elevation: float
    source: Literal["FLAT", "DEM_UPLOAD", "DEM_API", "DEM_SRTM", "API"]


class PhotoMetadataItem(BaseModel):
    """position metadata extracted from one uploaded photo - read-only, never persisted.

    coordinates carry [lon, lat, msl_alt] when the photo is geotagged; null when no GPS
    data was found (the batch does not fail). lens-height fields mirror the PAPI optics:
    msl is the raw absolute altitude, agl is msl minus DEM terrain (null when no DEM).
    """

    filename: str
    coordinates: PointZ | None = None
    lens_height_msl_m: float | None = None
    lens_height_agl_m: float | None = None
    error: str | None = None


class PhotoMetadataResponse(BaseModel):
    """batch response for photo position-metadata extraction."""

    items: list[PhotoMetadataItem]
    # whether a DEM backs the airport - when false, lens_height_agl_m is always null
    # and the frontend prompts for manual AGL entry
    has_dem: bool

File: frontend/src/types/airport.ts
import type { LineStringZ, PointZ, PolygonZ } from "./common";
import type {
  LampType,
  ObstacleType,
  PAPISide,
  SafetyZoneType,
  SurfaceType,
  TerrainSource,
} from "./enums";

export interface AirportResponse {
  id: string;
  icao_code: string;
  name: string;
  city: string | null;
  country: string | null;
  elevation: number;
  location: PointZ;
  default_drone_profile_id: string | null;
  terrain_source: TerrainSource;
  has_dem: boolean;
}

export interface AirportSummaryResponse extends AirportResponse {
  surfaces_count: number;
  agls_count: number;
  missions_count: number;
}

export interface AirportDetailResponse extends AirportResponse {
  surfaces: SurfaceResponse[];
  obstacles: ObstacleResponse[];
  safety_zones: SafetyZoneResponse[];
}

export interface SurfaceResponse {
  id: string;
  airport_id: string;
  identifier: string;
  surface_type: SurfaceType;
  geometry: LineStringZ;
  boundary: PolygonZ | null;
  buffer_distance: number;
  heading: number | null;
  length: number | null;
  width: number | null;
  threshold_position: PointZ | null;
  end_position: PointZ | null;
  touchpoint_latitude: number | null;
  touchpoint_longitude: number | null;
  touchpoint_altitude: number | null;
  paired_surface_id: string | null;
  agls: AGLResponse[];
}

export interface SurfaceCoupleRequest {
  target_surface_id: string;
  primary: "self" | "target";
}

export interface SurfaceCreateReverseRequest {
  identifier?: string | null;
}

export type AglType = "PAPI" | "RUNWAY_EDGE_LIGHTS";

export interface ObstacleResponse {
  id: string;
  airport_id: string;
  name: string;
  height: number;
  boundary: PolygonZ;
  buffer_distance: number;
  type: ObstacleType;
  position?: PointZ;
  radius?: number;
}

export interface SafetyZoneResponse {
  id: string;
  airport_id: string;
  name: string;
  type: SafetyZoneType;
  geometry: PolygonZ;
  altitude_floor: number | null;
  altitude_ceiling: number | null;
  is_active: boolean;
}

export interface AGLResponse {
  id: string;
  surface_id: string;
  agl_type: AglType;
  name: string;
  position: PointZ;
  side: PAPISide | null;
  glide_slope_angle: number | null;
  glide_slope_angle_tolerance: number | null;
  distance_from_threshold: number | null;
  meht_height_m: number | null;
  offset_from_centerline: number | null;
  lhas: LHAResponse[];
}

export interface LHAResponse {
  id: string;
  agl_id: string;
  unit_designator: string;
  // null for PAPI lhas before coordinator fills the angle in manually
  setting_angle: number | null;
  transition_sector_width: number | null;
  lamp_type: LampType;
  position: PointZ;
  tolerance: number | null;
  sequence_number: number;
  // PAPI-only lens height; null for non-PAPI units. msl is the raw absolute
  // altitude, agl is msl minus terrain elevation.
  lens_height_msl_m: number | null;
  lens_height_agl_m: number | null;
}

export interface AirportCreate {
  icao_code: string;
  name: string;
  city?: string | null;
  country?: string | null;
  elevation: number;
  location: PointZ;
}

export interface AirportUpdate {
  name?: string;
  city?: string | null;
  country?: string | null;
  elevation?: number;
  location?: PointZ;
}

export interface BulkChangeDroneResponse {
  updated_count: number;
  regressed_count: number;
  mission_ids: string[];
}

export interface TerrainCoverage {
  bounds: [number, number, number, number];
  resolution: [number, number];
}

export interface TerrainUploadResponse {
  terrain_source: TerrainSource;
  coverage: TerrainCoverage;
}

export interface TerrainDownloadResponse {
  terrain_source: TerrainSource;
  points_downloaded: number;
  coverage: TerrainCoverage;
}

export type ElevationSource = "FLAT" | "DEM_UPLOAD" | "DEM_API" | "API_FALLBACK";

export interface ElevationAtPointResponse {
  elevation: number;
  source: ElevationSource;
}

export interface SurfaceCreate {
  identifier: string;
  surface_type: SurfaceType;
  geometry: LineStringZ;
  boundary?: PolygonZ;
  buffer_distance?: number;
  heading?: number | null;
  length?: number | null;
  width?: number | null;
  threshold_position?: PointZ | null;
  end_position?: PointZ | null;
  touchpoint_latitude?: number | null;
  touchpoint_longitude?: number | null;
  touchpoint_altitude?: number | null;
}

export interface SurfaceUpdate {
  identifier?: string;
  geometry?: LineStringZ;
  // null clears the polygon boundary on the backend; undefined leaves it untouched
  boundary?: PolygonZ | null;
  buffer_distance?: number;
  heading?: number | null;
  length?: number | null;
  width?: number | null;
  threshold_position?: PointZ | null;
  end_position?: PointZ | null;
  touchpoint_latitude?: number | null;
  touchpoint_longitude?: number | null;
  touchpoint_altitude?: number | null;
}

export interface LHABulkGenerateRequest {
  first_position: PointZ;
  last_position: PointZ;
  spacing_m: number;
  setting_angle?: number | null;
  tolerance?: number | null;
  lamp_type?: "HALOGEN" | "LED";
}

export interface LHABulkGenerateResponse {
  generated: LHAResponse[];
}

export interface ObstacleCreate {
  name: string;
  height: number;
  boundary: PolygonZ;
  buffer_distance?: number;
  type: ObstacleType;
}

export interface ObstacleUpdate {
  name?: string;
  height?: number;
  boundary?: PolygonZ;
  buffer_distance?: number;
  type?: ObstacleType;
  preserve_altitude?: boolean;
}

export interface SafetyZoneCreate {
  name: string;
  type: SafetyZoneType;
  geometry: PolygonZ;
  altitude_floor?: number | null;
  altitude_ceiling?: number | null;
  is_active?: boolean;
}

export interface SafetyZoneUpdate {
  name?: string;
  type?: SafetyZoneType;
  geometry?: PolygonZ;
  altitude_floor?: number | null;
  altitude_ceiling?: number | null;
  is_active?: boolean;
}

export interface AGLCreate {
  agl_type: AglType;
  name: string;
  position: PointZ;
  side?: PAPISide | null;
  glide_slope_angle?: number | null;
  glide_slope_angle_tolerance?: number | null;
  distance_from_threshold?: number | null;
  meht_height_m?: number | null;
  offset_from_centerline?: number | null;
}

export interface AGLUpdate {
  agl_type?: AglType;
  name?: string;
  position?: PointZ;
  side?: PAPISide | null;
  glide_slope_angle?: number | null;
  glide_slope_angle_tolerance?: number | null;
  distance_from_threshold?: number | null;
  meht_height_m?: number | null;
  offset_from_centerline?: number | null;
  preserve_altitude?: boolean;
}

export interface LHACreate {
  unit_designator: string;
  setting_angle: number | null;
  transition_sector_width?: number | null;
  lamp_type: LampType;
  position: PointZ;
  tolerance?: number | null;
  sequence_number?: number | null;
  lens_height_msl_m?: number | null;
  lens_height_agl_m?: number | null;
}

export interface LHAUpdate {
  unit_designator?: string;
  setting_angle?: number | null;
  transition_sector_width?: number | null;
  lamp_type?: LampType;
  position?: PointZ;
  tolerance?: number | null;
  sequence_number?: number;
  preserve_altitude?: boolean;
  lens_height_msl_m?: number | null;
  lens_height_agl_m?: number | null;
}

// position metadata extracted from one uploaded photo (read-only, never
// persisted). coordinates carry [lon, lat, msl_alt] when geotagged, null when
// no GPS data was found. lens-height fields mirror the PAPI optics.
export interface PhotoMetadataItem {
  filename: string;
  coordinates: PointZ | null;
  lens_height_msl_m: number | null;
  lens_height_agl_m: number | null;
  error: string | null;
}

export interface PhotoMetadataResponse {
  items: PhotoMetadataItem[];
  // when false, lens_height_agl_m is always null and AGL is manual entry
  has_dem: boolean;
}

export interface SurfaceDimensions {
  length: number | null;
  width: number | null;
  heading: number | null;
}

export interface SurfaceRecalculateResponse {
  current: SurfaceDimensions;
  recalculated: SurfaceDimensions;
}

export interface ObstacleDimensions {
  length: number | null;
  width: number | null;
  heading: number | null;
  radius: number | null;
}

export interface ObstacleRecalculateResponse {
  current: ObstacleDimensions;
  recalculated: ObstacleDimensions;
}

// openaip lookup suggestions
export interface RunwaySuggestion {
  identifier: string;
  heading: number;
  length: number;
  width: number;
  threshold_position: PointZ;
  end_position: PointZ;
  geometry: LineStringZ;
  boundary: PolygonZ;
}

export interface ObstacleSuggestion {
  name: string;
  type: ObstacleType;
  height: number;
  boundary: PolygonZ;
}

export interface SafetyZoneSuggestion {
  name: string;
  type: SafetyZoneType;
  geometry: PolygonZ;
  altitude_floor: number | null;
  altitude_ceiling: number | null;
}

export interface AirportLookupResponse {
  icao_code: string;
  name: string;
  city: string | null;
  country: string | null;
  elevation: number;
  location: PointZ;
  runways: RunwaySuggestion[];
  obstacles: ObstacleSuggestion[];
  safety_zones: SafetyZoneSuggestion[];
}
